### ISSUE TITLE
Script ending when Steam instalation window appears

### ISSUE THREAD

**Issue by Zemogiter** [PhoenicisOrg/scripts#627]
I was trying to install Subnautica but the second the script started Steam instalation it crashed with this error log:

```
java.lang.IllegalMonitorStateException
	at java.lang.Object.wait(Native Method)
	at jdk.nashorn.internal.scripts.Script$Recompilation$286$2195$\^eval\_.go([engines, wine, quick_script, steam_script]:103)
	at jdk.nashorn.internal.scripts.Script$Recompilation$272$684$\^eval\_.run(<eval>:18)
	at org.phoenicis.scripts.Installer$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$239$\^eval\_.:program(<eval>:47)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by plata** [comment]
Can you try:

```Javascript
 setupWizard.wait(tr("Please follow the steps of the Steam setup.\n\nUncheck \"Run Steam\" or close Steam completely after the setup so that the installation of \"{0}\" can continue.", this._name));

var wine = new Wine()
        .wizard(setupWizard)
        .prefix(this._name, this._wineDistribution, this._wineArchitecture, this._wineVersion)
        .luna()
        .run(tempFile, [], wine.prefixDirectory(), false, true);
```

**Comment by Zemogiter** [comment]
I did and now it gives an error about not recognizing prefixDirectory
``
avax.script.ScriptException: TypeError: Cannot read property "prefixDirectory" from undefined in [engines, wine, quick_script, steam_script] at line number 103
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: [engines, wine, quick_script, steam_script]:103 TypeError: Cannot read property "prefixDirectory" from undefined
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:213)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:185)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:172)
	at jdk.nashorn.internal.runtime.Undefined.get(Undefined.java:157)
	at jdk.nashorn.internal.scripts.Script$Recompilation$564$2195$\^eval\_.go([engines, wine, quick_script, steam_script]:103)
	at jdk.nashorn.internal.scripts.Script$Recompilation$550$684$\^eval\_.run(<eval>:18)
	at org.phoenicis.scripts.Installer$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$517$\^eval\_.:program(<eval>:47)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
``

**Comment by plata** [comment]
The first error is because the `Wine` object is just being created. Should be possible to work around this with:
```Javascript
setupWizard.wait(tr("Please follow the steps of the Steam setup.\n\nUncheck \"Run Steam\" or close Steam completely after the setup so that the installation of \"{0}\" can continue.", this._name));

var wine = new Wine()
        .wizard(setupWizard)
        .prefix(this._name, this._wineDistribution, this._wineArchitecture, this._wineVersion)
        .luna()
        .run(tempFile, [], null, false, true);
```

**Comment by Zemogiter** [comment]
Now the script finishes installing Steam but once that is done it crashes with wine.vcrun2017 is not a function in <eval> at line number 25.

**Comment by plata** [comment]
For `vcrun2017`: which application is affected? Does it include the `vcrun2017` verb?

**Comment by Zemogiter** [comment]
Two applications are affected: Space Engineers (haven't uploaded script for that yet because there are some dotnet-related bugs in wine that prevent this game from working) and Subnautica. The former requires vcrun2017 while the latter needs vcrun2013 but the error is the same. And both scripts have their verbs included.

**Comment by plata** [comment]
I've already done the Uplay update (just copied the changes from your pull request).

Currently I have no idea for the verbs. I will test it locally.

**Comment by Zemogiter** [comment]
The same error is displayed for vcrun2008 which was mentioned in the latest test result at WineHQ as a requriment for Steam.

**Comment by Zemogiter** [comment]
> I'm really sorry for the trouble but we're reworking the engines to make them more generic and it will probably take some time until that really stabilized.

Yeah I'm aware of that since you and @madoar and @ImperatorS79 were talking about that for some time in issues and pull requests. To check if it was Steam/Uplay specific bug I've tested an local install script for Wildlife Park 2 and aside from quartz progress bar not showing up (after the dialoge and clicking "Next" the script appears to be stuck untill quartz is installed, checked the quartz verb and its fine) the instalation went fine. But when I wanted to play the game it crashed after the intro. No such thing on POL4 or vanilla Wine. There is nothing in phoenicis.log that would point me to anything specific causing this problem so I can't create a new issue.

**Comment by plata** [comment]
Can we please separate this Wildlife Park 2 issue? I would suggest to wait until my PR is merge and then test again.

**Comment by Zemogiter** [comment]
Sure

**Comment by plata** [comment]
The PR is merged. Feel free to test again with current master.

**Comment by Zemogiter** [comment]
I've tested after refreshing repos and removing the local one. Now I'm getting this error:
`javax.script.ScriptException: TypeError: fileUtilities.ls is not a function in [utils, functions, filesystem, files] at line number 11
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: [utils, functions, filesystem, files]:11 TypeError: fileUtilities.ls is not a function
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:213)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:185)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:172)
	at jdk.nashorn.internal.runtime.Undefined.lookup(Undefined.java:102)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:106)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:98)
	at jdk.internal.dynalink.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(CompositeTypeBasedGuardingDynamicLinker.java:176)
	at jdk.internal.dynalink.support.CompositeGuardingDynamicLinker.getGuardedInvocation(CompositeGuardingDynamicLinker.java:124)
	at jdk.internal.dynalink.support.LinkerServicesImpl.getGuardedInvocation(LinkerServicesImpl.java:154)
	at jdk.internal.dynalink.DynamicLinker.relink(DynamicLinker.java:253)
	at jdk.nashorn.internal.scripts.Script$Recompilation$492$300A$\^eval\_.ls([utils, functions, filesystem, files]:11)
	at jdk.nashorn.internal.scripts.Script$Recompilation$485$3776AAA$\^eval\_._installWinePackage([engines, wine, engine, implementation]:98)
	at jdk.nashorn.internal.scripts.Script$Recompilation$484$3158ADA$\^eval\_.install#L:63#L:65([engines, wine, engine, implementation]:67)
	at jdk.nashorn.internal.objects.NativeArray$9.forEach(NativeArray.java:1615)
	at jdk.nashorn.internal.runtime.arrays.IteratorAction.apply(IteratorAction.java:110)
	at jdk.nashorn.internal.objects.NativeArray.forEach(NativeArray.java:1618)
	at jdk.nashorn.internal.scripts.Script$Recompilation$483$3027ADA$\^eval\_.install#L:63([engines, wine, engine, implementation]:65)
	at jdk.nashorn.internal.objects.NativeArray$9.forEach(NativeArray.java:1615)
	at jdk.nashorn.internal.runtime.arrays.IteratorAction.apply(IteratorAction.java:110)
	at jdk.nashorn.internal.objects.NativeArray.forEach(NativeArray.java:1618)
	at jdk.nashorn.internal.scripts.Script$Recompilation$412$2224AA$\^eval\_.install([engines, wine, engine, implementation]:63)
	at jdk.nashorn.internal.scripts.Script$Recompilation$414$7731AAAAAA$\^eval\_.run([engines, wine, engine, implementation]:198)
	at org.phoenicis.engines.Engine$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$Recompilation$479$4764AAAZZA$\^eval\_.run([engines, wine, engine, object]:168)
	at jdk.nashorn.internal.scripts.Script$Recompilation$478$251A$\^eval\_.regedit#open([engines, wine, plugins, regedit]:12)
	at jdk.nashorn.internal.scripts.Script$Recompilation$467$348$\^eval\_.luna([engines, wine, verbs, luna]:29)
	at jdk.nashorn.internal.scripts.Script$Recompilation$446$2195$\^eval\_.go([engines, wine, quick_script, steam_script]:100)
	at jdk.nashorn.internal.scripts.Script$Recompilation$432$684$\^eval\_.run(<eval>:18)
	at org.phoenicis.scripts.Installer$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$399$\^eval\_.:program(<eval>:47)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
`

**Comment by plata** [comment]
Is your phoenicis software up-to-date? (that feature is quite new but I have definitely used it).

**Comment by Zemogiter** [comment]
I have updated POL5 and made a local copy of the repo. This is the error it gives now:
```
include(["engines", "wine", "quick_script", "quick_script"]);
include(["utils", "functions", "net", "download"]);
include(["engines", "wine", "engine", "object"]);
include(["engines", "wine", "plugins", "override_dll"]);
include(["utils", "functions", "filesystem", "extract"]);
include(["utils", "functions", "filesystem", "files"]);
include(["engines", "wine", "verbs", "luna"]);


function SteamScript() {
    QuickScript.call(this);

    this._executable = "Steam.exe";
    this._category = "Games";
    this._gameOverlay = true;
}

SteamScript.prototype = Object.create(QuickScript.prototype);

SteamScript.prototype.constructor = SteamScript;

SteamScript.prototype.appId = function (appId) {
    this._appId = appId;
    return this;
};

SteamScript.prototype.gameOverlay = function (gameOverlay) {
    // get
    if (arguments.length == 0) {
        return this._gameOverlay;
    }

    // set
    this._gameOverlay = gameOverlay;
    return this;
};

SteamScript.prototype.manifest = function (wine) {
    if (!this._manifest) {
        // cache manifest path (will not change during the installation)
        this._manifest = wine.prefixDirectory() + "/drive_c/" + wine.programFiles() + "/Steam/steamapps/appmanifest_" + this._appId + ".acf";
    }
    return this._manifest;
};

SteamScript.prototype.downloadStarted = function (wine) {
    if (fileExists(this.manifest(wine)))
    {
        var manifest = cat(this.manifest(wine));
        var state = Number(manifest.match(/\"StateFlags\"\s+\"(\d+)\"/)[1]);
        return state == 1026 || state == 1042 || state == 1062 || state == 1030;
    }
    else
    {
        return false;
    }
};

SteamScript.prototype.downloadFinished = function (wine) {
    // check if download already finished (download folder has been deleted)
    if (fileExists(this.manifest(wine)))
    {
        var manifest = cat(this.manifest(wine));
        var state = Number(manifest.match(/\"StateFlags\"\s+\"(\d+)\"/)[1]);
        return state != 1026 && state != 1042 && state != 1062 && state != 1030;
    }
    else
    {
        return false;
    }
};

SteamScript.prototype.go = function () {
    // default application homepage if not specified
    if (!this._applicationHomepage) {
        this._applicationHomepage = "http://store.steampowered.com/app/" + this._appId;
    }

    // default executable args if not specified
    if (!this._executableArgs) {
        this._executableArgs = ["-no-cef-sandbox", "-silent", "-applaunch", this._appId];
    }

    var setupWizard = SetupWizard(InstallationType.APPS, this._name, this.miniature());

    setupWizard.presentation(this._name, this._editor, this._applicationHomepage, this._author);

    var tempFile = createTempFile("exe");

    new Downloader()
        .wizard(setupWizard)
        .url("http://media.steampowered.com/client/installer/SteamSetup.exe")
        .checksum("4b1b85ec2499a4ce07c89609b256923a4fc479e5")
        .to(tempFile)
        .get();

    setupWizard.wait(tr("Please follow the steps of the Steam setup.\n\nUncheck \"Run Steam\" or close Steam completely after the setup so that the installation of \"{0}\" can continue.", this._name));

    var wine = new Wine()
        .wizard(setupWizard)
        .prefix(this._name, this._wineDistribution, this._wineArchitecture, this._wineVersion)
        .luna()
        .run(tempFile, [], null, false, true);

    // Steam installation has finished
    setupWizard.wait(tr("Please wait ..."));

    this._preInstall(wine, setupWizard);

    // back to generic wait (might have been changed in preInstall)
    setupWizard.wait(tr("Please wait ..."));

    wine.runInsidePrefix(wine.programFiles() + "/Steam/Steam.exe", ["steam://install/" + this._appId]);

    setupWizard.wait(tr("Please wait until Steam has finished the download ..."));

    // wait until download started
    while (!this.downloadStarted(wine)) {
        java.lang.Thread.sleep(100);
    }

    // make sure download is finished
    while (!this.downloadFinished(wine)) {
        java.lang.Thread.sleep(1000);
    }

    // close Steam
    wine.runInsidePrefix(wine.programFiles() + "/Steam/Steam.exe", "-shutdown");

    // back to generic wait
    setupWizard.wait(tr("Please wait ..."));

    // create shortcut after installation (if executable is specified, it does not exist earlier)
    this._createShortcut(wine.prefix());

    this._postInstall(wine, setupWizard);

    // deactivate game overlay if desired
    if (!this._gameOverlay) {
        wine.overrideDLL()
            .set("", ["gameoverlayrenderer"])
            .do();
    }

    // back to generic wait (might have been changed in postInstall)
    setupWizard.wait(tr("Please wait ..."));

    setupWizard.close();
};
```

This line seems to be the problem:
wine.runInsidePrefix(wine.programFiles() + "/Steam/Steam.exe", ["steam://install/" + this._appId]);

**Comment by plata** [comment]
Would you mind providing the fix for the other files as well? You don't have to test it.
